### PR TITLE
ORM: Fix typing of `aiida.orm.nodes.data.code` module

### LINKED ISSUE #5819: Fix type errors in orm.code modules
If the orm code modules are removed from the pre-commit mypy excluded list, below are the current errors.

1. Issues assuming that a computer has been set
2. Issues about what type is used for the executable path (see also https://github.com/aiidateam/aiida-core/issues/5797#issuecomment-1332433139)

It would be good to get these fixed

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

aiida/orm/nodes/data/code/portable.py:57: error: Incompatible types in assignment (expression has type "str", variable has type "PurePath")  [assignment]
aiida/orm/nodes/data/code/portable.py:58: error: Argument 1 to "put_object_from_tree" of "NodeRepository" has incompatible type "Path"; expected "str"  [arg-type]
aiida/orm/nodes/data/code/portable.py:94: error: Incompatible return value type (got "PurePath", expected "Path")  [return-value]
aiida/orm/nodes/data/code/installed.py:45: error: Incompatible types in assignment (expression has type "str", variable has type "PurePath")  [assignment]
aiida/orm/nodes/data/code/installed.py:74: error: Item "None" of "Optional[Computer]" has no attribute "get_transport"  [union-attr]
aiida/orm/nodes/data/code/installed.py:93: error: Item "None" of "Optional[Computer]" has no attribute "pk"  [union-attr]
aiida/orm/nodes/data/code/installed.py:100: error: Incompatible return value type (got "PurePath", expected "Path")  [return-value]
aiida/orm/nodes/data/code/installed.py:111: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:144: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:152: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:160: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:169: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:208: error: Item "List[Any]" of "Union[List[Any], Any]" has no attribute "pk"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:212: error: Value of type "Optional[List[Any]]" is not indexable  [index]
aiida/orm/nodes/data/code/legacy.py:306: error: Item "List[Any]" of "Union[List[Any], Any]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:308: error: Item "List[Any]" of "Union[List[Any], Any]" has no attribute "pk"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:352: error: Item "None" of "Optional[Computer]" has no attribute "get_transport"  [union-attr]
Found 17 errors in 3 files (checked 728 source files)
```

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #5819

### CODE REVIEW

**Comment by chrisjsewell** [comment]
Heya, so as I've mentioned before, the problem with using `pathlib.PurePath` here, is that it implies that one can also represent paths on Windows (aka DOS) systems, and I don't believe that is the case.

For starters, I don't think `SshTransport` is compatible with the remote computer being a Windows machine anyway.

Then for the code:

1. Since aiida can only be installed on UNIX, this means that `pathlib.PurePath` will actually always be `pathlib.PurePosixPath`.
2. Since the path is serialised to the storage as a string, even if the user would specifically add the executable path as `PureWindowsPath`, it would still end up being read as a unix path, e.g. `PurePath(str(PureWindowsPath(r'C:\path\to\exec.exe'))).parent` -> `PurePosixPath('.')` which is obviously wrong.
   - if you actually wantd to allow for windows paths, then you would need to specifically serialise the path somehow with specific information on the OS

So, basically, I think you should change all the `PurePath` uses to `PurePosixPath`

**Comment by sphuber** [comment]
But what you are describing is a limitation of the `Transport` interface and the limitations other parts of AiiDA impose on the remote computer. But in principle the `Code` is abstract in that sense and should be able to support both no? It is possible that we implement support in the future for running on remote windows computers, and in that case the code classes should just work with `PurePath`. Is there any downside to keeping `PurePath` now?

**Comment by chrisjsewell** [comment]
The `Code.executable_path` is currently stored in the database as a string.
How do you know whether that string is intended to be posix or windows?
Currently it is essentially interpreted as the same as the filesystem that aiida is installed on, which is wrong.

**Comment by chrisjsewell** [comment]
At present the problem is, by wrapping things in `pathlib` objects you have essentially made the `Code` less abstract than it previously was.

**Comment by sphuber** [comment]
> Currently it is essentially interpreted as the same as the filesystem that aiida is installed on, which is wrong.

Ah I see now, yeah you are right, that is a problem.

**Review by chrisjsewell — APPROVED** [review]
cheers
